src/lib.rs

```diff
@@ -16,116 +16,261 @@
-    if major == 3 && minor < 1 || major < 3 {
-        panic!(
-            "Invalid version of protoc (required at least 3.1.x, get {}.{}.x).",
-            major, minor,
-        );
+/// Generate Rust files from proto files (`files`).
+pub fn generate_files<T: AsRef<Path> + Debug>(includes: &[T], files: &[T], out_dir: &str) {
+    if Path::new(&out_dir).exists() {
+        fs::remove_dir_all(&out_dir).unwrap();
+    }
+    fs::create_dir_all(&out_dir).unwrap();
+    generate(includes, files, out_dir);
+    let mut generated = Vec::new();
+    let modules: Vec<_> = fs::read_dir(out_dir)
+        .unwrap()
+        .filter_map(|res| {
+            let path = match res {
+                Ok(e) => e.path(),
+                Err(e) => panic!("failed to list {}: {:?}", out_dir, e),
+            };
+            if path.extension() == Some(std::ffi::OsStr::new("rs")) {
+                generated.push(format!("{}", path.display()));
+                let name = path.file_stem().unwrap().to_str().unwrap();
+                Some((name.replace('-', "_"), name.to_owned()))
+            } else {
+                None
+            }
+        })
+        .collect();
+    replace_read_unknown_fields(&generated);
+    let mut f = File::create(format!("{}/mod.rs", out_dir)).unwrap();
+    for (module, file_name) in &modules {
+        if !module.contains('.') {
+            writeln!(f, "pub mod {};", module).unwrap();
+            continue;
+        }
+        let mut level = 0;
+        for part in module.split('.') {
+            writeln!(f, "{:level$}pub mod {} {{", "", part, level = level).unwrap();
+            level += 1;
+        }
+        writeln!(f, "include!(\"{}.rs\");", file_name).unwrap();
+        for _ in (0..level).rev() {
+            writeln!(f, "{:1$}}}", "", level).unwrap();
+        }
     }
 }
 
-/// Use protobuf-rs to generate Rust files from proto files (`file_names`).
-///
-/// Uses `["proto", "include"]` as the include lists.
-pub fn generate_protobuf_files<T: AsRef<str>>(file_names: &[T], out_dir: &str) {
-    let file_names = &file_names.iter().map(AsRef::as_ref).collect::<Vec<_>>();
-    protoc_rust::run(protoc_rust::Args {
-        out_dir,
-        input: file_names,
-        includes: &["proto", "include"],
-        customize: protoc_rust::Customize {
-            ..Default::default()
-        },
-    })
-    .unwrap();
-
-    protoc_grpcio::compile_grpc_protos(file_names, &["proto", "include"], out_dir, None).unwrap();
+/// Use rust-protobuf to generate Rust files from proto files (`files`).
+#[cfg(feature = "protobuf-codec")]
+mod protobuf_imps {
+    use regex::Regex;
+    use std::env;
+    use std::fmt::Debug;
+    use std::path::Path;
+    use std::process::Command;
+    use std::str::from_utf8;
+
+    pub fn get_protoc() -> String {
+        let protoc_bin_name = match (env::consts::OS, env::consts::ARCH) {
+            ("linux", "x86") => "protoc-linux-x86_32",
```

Comment:
I think we do not need x86 or ppc support - I don't think anyone will use these and it just adds to the download size.